patches/net/minecraft/world/level/block/state/BlockBehaviour.java.patch

```diff
@@ -150,6 +165,22 @@
          BlockBehaviour.StatePredicate isRedstoneConductor = (p_284888_, p_284889_, p_284890_) -> p_284888_.isCollisionShapeFullBlock(p_284889_, p_284890_);
          BlockBehaviour.StatePredicate isSuffocating = (p_284885_, p_284886_, p_284887_) -> p_284885_.blocksMotion()
                  && p_284885_.isCollisionShapeFullBlock(p_284886_, p_284887_);
+@@ -974,6 +_,7 @@
+         FeatureFlagSet requiredFeatures = FeatureFlags.VANILLA_SET;
+         @Nullable
+         BlockBehaviour.OffsetFunction offsetFunction;
++        private final it.unimi.dsi.fastutil.objects.ReferenceSet<net.neoforged.neoforge.flag.Flag> requiredFlags = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
```

Comment:
Even if we use fastutil reference sets, my point from before still stands; the `requiredFlags` fields should be implemented in such a way that we are not storing a bunch of empty `ReferenceOpenHashSet` objects for blocks without any required flags. The memory usage will add up quickly in situations where there are a lot of single-state blocks (which is common with building block mods, for instance).

Reply:
Makes sense, if no flags are defined why store a empty set for every block (all of vanilla would be have said empty sets).

I could update to have a global immutable empty set stored statically in `FlagManager` which everything delegates to if no other flags could be found/are defined.

Reply:
The simplest solution would be to make the field non-final and either initialize it to `ReferenceSets.emptySet()` and only replace it with a mutable set the first time a flag is added or patch the block constructor to replace the mutable set with `ReferenceSets.emptySet()` if it's empty. The former will likely be easier to use in other places, especially ones were no builder-like structure is involved. It's also worth noting that blocks are a particularly annoying case because they keep around the `BlockBehaviour.Properties` object for their entire lifetime. This means that you need to discard the mutable set on both the block and its properties if it's empty and also avoid copying the flags set when transferring it to the block.

Reply:
Since blocks keep the properties around for their whole lifce cycle, for them specificly would it not be better to delegate to the set within the properties rather than constructing and storing a new one in the block itself?

Everything else i can update to match as xfact suggested, empty by default but once a flag is defined its updated to a mutable set

Reply:
Sure, keeping it only on the properties also works. Make sure to update the properties copying to take the initial empty set into account.